File: synesis/auth/oauth.py
from __future__ import annotations

import hashlib
import secrets
import subprocess
import sys
import time
from http.server import HTTPServer, BaseHTTPRequestHandler
from threading import Thread
from urllib.parse import urlencode, urlparse, parse_qs

import httpx

from synesis.auth.store import AuthStore

CALLBACK_PORT = 9876


class OAuthManager:
    def __init__(self, project_dir: str):
        self.store = AuthStore(project_dir)

    def init(self) -> None:
        self.store.init()

    def authenticate(self, provider: dict) -> dict | None:
        # Check for existing valid tokens
        existing = self.store.load(provider["name"])
        if existing:
            expires_at = existing.get("expires_at", 0)
            if expires_at > time.time() + 300:
                return existing
            if existing.get("refresh_token"):
                try:
                    refreshed = self._refresh_token(provider, existing["refresh_token"])
                    self.store.save(provider["name"], refreshed)
                    return refreshed
                except Exception:
                    pass

        # Start OAuth flow
        tokens = self._run_oauth_flow(provider)
        if tokens:
            self.store.save(provider["name"], tokens)
        return tokens

    def get_token(self, provider_name: str) -> dict | None:
        return self.store.load(provider_name)

    def list_authenticated(self) -> list[str]:
        return self.store.list()

    def revoke(self, provider_name: str) -> bool:
        return self.store.delete(provider_name)

    def _run_oauth_flow(self, provider: dict) -> dict | None:
        state = secrets.token_hex(16)
        code_verifier = secrets.token_urlsafe(32)
        code_challenge = (
            hashlib.sha256(code_verifier.encode())
            .digest()
            .hex()
        )
        # Proper base64url encoding for PKCE
        import base64
        code_challenge = base64.urlsafe_b64encode(
            hashlib.sha256(code_verifier.encode()).digest()
        ).rstrip(b"=").decode()

        callback_url = f"http://localhost:{CALLBACK_PORT}/callback"

        params = {
            "client_id": provider["client_id"],
            "redirect_uri": callback_url,
            "response_type": "code",
            "scope": " ".join(provider["scopes"]),
            "state": state,
            "code_challenge": code_challenge,
            "code_challenge_method": "S256",
            "access_type": "offline",
            "prompt": "consent",
        }

        auth_url = f"{provider['auth_url']}?{urlencode(params)}"
        print(f"\nOpen this URL to authenticate with {provider['name']}:\n")
        print(auth_url)
        print("\nWaiting for callback...")

        # Try to open browser
        try:
            if sys.platform == "darwin":
                subprocess.Popen(["open", auth_url])
            elif sys.platform == "linux":
                subprocess.Popen(["xdg-open", auth_url])
        except Exception:
            pass

        # Wait for callback
        code = self._wait_for_callback(state)
        if not code:
            return None

        return self._exchange_code(provider, code, callback_url, code_verifier)

    def _wait_for_callback(self, expected_state: str) -> str | None:
        result = {"code": None}

        class Handler(BaseHTTPRequestHandler):
            def do_GET(self):
                parsed = urlparse(self.path)
                params = parse_qs(parsed.query)

                if parsed.path != "/callback":
                    self.send_response(404)
                    self.end_headers()
                    return

                error = params.get("error", [None])[0]
                if error:
                    self.send_response(400)
                    self.end_headers()
                    self.wfile.write(f"Error: {error}".encode())
                    return

                state = params.get("state", [None])[0]
                if state != expected_state:
                    self.send_response(400)
                    self.end_headers()
                    self.wfile.write(b"Invalid state")
                    return

                code = params.get("code", [None])[0]
                if not code:
                    self.send_response(400)
                    self.end_headers()
                    return

                result["code"] = code
                self.send_response(200)
                self.send_header("Content-Type", "text/html")
                self.end_headers()
                self.wfile.write(
                    b"<html><body style='font-family:system-ui;display:flex;justify-content:center;"
                    b"align-items:center;height:100vh;margin:0'>"
                    b"<div style='text-align:center'><h1>Synesis</h1>"
                    b"<p>Authentication successful. You can close this tab.</p></div>"
                    b"</body></html>"
                )

            def log_message(self, *args):
                pass  # Suppress logs

        server = HTTPServer(("localhost", CALLBACK_PORT), Handler)
        server.timeout = 120

        # Run in thread with timeout
        thread = Thread(target=lambda: server.handle_request())
        thread.start()
        thread.join(timeout=120)

        server.server_close()
        return result["code"]

    def _exchange_code(
        self, provider: dict, code: str, redirect_uri: str, code_verifier: str
    ) -> dict | None:
        with httpx.Client() as client:
            resp = client.post(
                provider["token_url"],
                data={
                    "grant_type": "authorization_code",
                    "code": code,
                    "redirect_uri": redirect_uri,
                    "client_id": provider["client_id"],
                    "client_secret": provider["client_secret"],
                    "code_verifier": code_verifier,
                },
                headers={"Accept": "application/json"},
            )
            resp.raise_for_status()
            data = resp.json()

            return {
                "access_token": data["access_token"],
                "refresh_token": data.get("refresh_token"),
                "expires_at": time.time() + data.get("expires_in", 3600),
                "token_type": data.get("token_type", "Bearer"),
                "scope": data.get("scope"),
            }

    def _refresh_token(self, provider: dict, refresh_token: str) -> dict:
        with httpx.Client() as client:
            resp = client.post(
                provider["token_url"],
                data={
                    "grant_type": "refresh_token",
                    "refresh_token": refresh_token,
                    "client_id": provider["client_id"],
                    "client_secret": provider["client_secret"],
                },
                headers={"Accept": "application/json"},
            )
            resp.raise_for_status()
            data = resp.json()

            return {
                "access_token": data["access_token"],
                "refresh_token": data.get("refresh_token", refresh_token),
                "expires_at": time.time() + data.get("expires_in", 3600),
                "token_type": data.get("token_type", "Bearer"),
                "scope": data.get("scope"),
            }

File: src/auth/oauth.ts
import http from "http";
import { URL } from "url";
import crypto from "crypto";
import type { OAuthProvider, TokenSet } from "./types.js";
import { AuthStore } from "./store.js";

const CALLBACK_PORT = 9876;
const CALLBACK_HOST = "localhost";

export class OAuthManager {
  private store: AuthStore;

  constructor(projectDir: string) {
    this.store = new AuthStore(projectDir);
  }

  async init(): Promise<void> {
    await this.store.init();
  }

  async authenticate(provider: OAuthProvider): Promise<TokenSet> {
    // Check for existing valid tokens
    const existing = await this.store.load(provider.name);
    if (existing && existing.tokens.expires_at) {
      if (existing.tokens.expires_at > Date.now() / 1000 + 300) {
        return existing.tokens;
      }
      // Try refresh
      if (existing.tokens.refresh_token) {
        try {
          const refreshed = await this.refreshToken(provider, existing.tokens.refresh_token);
          await this.store.save(provider.name, refreshed);
          return refreshed;
        } catch {
          // Fall through to re-auth
        }
      }
    }

    // Start OAuth flow
    const tokens = await this.startOAuthFlow(provider);
    await this.store.save(provider.name, tokens);
    return tokens;
  }

  async getToken(providerName: string): Promise<TokenSet | null> {
    const auth = await this.store.load(providerName);
    return auth?.tokens || null;
  }

  async listAuthenticated(): Promise<string[]> {
    return this.store.list();
  }

  async revoke(providerName: string): Promise<boolean> {
    return this.store.delete(providerName);
  }

  private async startOAuthFlow(provider: OAuthProvider): Promise<TokenSet> {
    const state = crypto.randomBytes(16).toString("hex");
    const codeVerifier = crypto.randomBytes(32).toString("base64url");
    const codeChallenge = crypto
      .createHash("sha256")
      .update(codeVerifier)
      .digest("base64url");

    const callbackUrl = `http://${CALLBACK_HOST}:${CALLBACK_PORT}${provider.callbackPath || "/callback"}`;

    const authParams = new URLSearchParams({
      client_id: provider.clientId,
      redirect_uri: callbackUrl,
      response_type: "code",
      scope: provider.scopes.join(" "),
      state,
      code_challenge: codeChallenge,
      code_challenge_method: "S256",
      access_type: "offline",
      prompt: "consent",
    });

    const authUrl = `${provider.authUrl}?${authParams.toString()}`;

    console.log(`\nOpen this URL in your browser to authenticate with ${provider.name}:\n`);
    console.log(authUrl);
    console.log(`\nWaiting for callback...`);

    // Try to open browser automatically
    const { exec } = await import("child_process");
    exec(`open "${authUrl}"`);

    // Start local server to receive callback
    const code = await this.waitForCallback(state, provider.callbackPath || "/callback");

    // Exchange code for tokens
    const tokens = await this.exchangeCode(provider, code, callbackUrl, codeVerifier);
    return tokens;
  }

  private waitForCallback(expectedState: string, callbackPath: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        server.close();
        reject(new Error("OAuth callback timed out (120s)"));
      }, 120000);

      const server = http.createServer((req, res) => {
        const url = new URL(req.url || "", `http://${CALLBACK_HOST}:${CALLBACK_PORT}`);

        if (url.pathname !== callbackPath) {
          res.writeHead(404);
          res.end("Not found");
          return;
        }

        const code = url.searchParams.get("code");
        const state = url.searchParams.get("state");
        const error = url.searchParams.get("error");

        if (error) {
          res.writeHead(400);
          res.end(`Authentication failed: ${error}`);
          clearTimeout(timeout);
          server.close();
          reject(new Error(`OAuth error: ${error}`));
          return;
        }

        if (state !== expectedState) {
          res.writeHead(400);
          res.end("Invalid state parameter");
          clearTimeout(timeout);
          server.close();
          reject(new Error("OAuth state mismatch"));
          return;
        }

        if (!code) {
          res.writeHead(400);
          res.end("No code received");
          clearTimeout(timeout);
          server.close();
          reject(new Error("No OAuth code received"));
          return;
        }

        res.writeHead(200, { "Content-Type": "text/html" });
        res.end(`
          <html><body style="font-family: system-ui; display: flex; justify-content: center; align-items: center; height: 100vh; margin: 0;">
            <div style="text-align: center;">
              <h1>Synesis</h1>
              <p>Authentication successful. You can close this tab.</p>
            </div>
          </body></html>
        `);

        clearTimeout(timeout);
        server.close();
        resolve(code);
      });

      server.listen(CALLBACK_PORT, CALLBACK_HOST, () => {
        // Server ready
      });
    });
  }

  private async exchangeCode(
    provider: OAuthProvider,
    code: string,
    redirectUri: string,
    codeVerifier: string
  ): Promise<TokenSet> {
    const body = new URLSearchParams({
      grant_type: "authorization_code",
      code,
      redirect_uri: redirectUri,
      client_id: provider.clientId,
      client_secret: provider.clientSecret,
      code_verifier: codeVerifier,
    });

    const response = await fetch(provider.tokenUrl, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: body.toString(),
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Token exchange failed: ${response.status} ${text}`);
    }

    const data = await response.json();
    return {
      access_token: data.access_token,
      refresh_token: data.refresh_token,
      expires_at: data.expires_in
        ? Math.floor(Date.now() / 1000) + data.expires_in
        : undefined,
      token_type: data.token_type || "Bearer",
      scope: data.scope,
    };
  }

  private async refreshToken(
    provider: OAuthProvider,
    refreshToken: string
  ): Promise<TokenSet> {
    const body = new URLSearchParams({
      grant_type: "refresh_token",
      refresh_token: refreshToken,
      client_id: provider.clientId,
      client_secret: provider.clientSecret,
    });

    const response = await fetch(provider.tokenUrl, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: body.toString(),
    });

    if (!response.ok) {
      throw new Error(`Token refresh failed: ${response.status}`);
    }

    const data = await response.json();
    return {
      access_token: data.access_token,
      refresh_token: data.refresh_token || refreshToken,
      expires_at: data.expires_in
        ? Math.floor(Date.now() / 1000) + data.expires_in
        : undefined,
      token_type: data.token_type || "Bearer",
      scope: data.scope,
    };
  }
}
